### PR TITLE
[REVIEW] FEA MST

### LINKED ISSUE #1088: [FEA] Expose Minimum/Maximum Spanning Tree/Forest (MST)
https://en.wikipedia.org/wiki/Minimum_spanning_tree

Backend described in RAFT : https://github.com/rapidsai/raft/issues/52
Expose as a cuGraph algorithm by adding bindings, doc and python tests for it

### PR DESCRIPTION
**Pull request by afender**
This is wrapping the core solver in RAFT https://github.com/rapidsai/raft/pull/101. 
I will connect to RAFT PR and launch CI once it is merged. This PR is independent and can be reviewed in parallel.

Closes #1088 

Features
- Minimum spanning Tree + Forest
- Maximum spanning Tree + Forest 

Boruvka algorithm with support for :
- negative edge weights
- equal weights
- multi graph

Extras
- Script for sparsity analysis
- Script for npz to mtx

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/include/graph.hpp [hunk 1]
```diff
@@ -378,15 +378,15 @@ class GraphCOO {
...
+  rmm::device_buffer edge_data_p{};    ///< data
 
  public:
   /**
    * @brief      Take ownership of the provided graph arrays in COO format
    *
    * @param  number_of_vertices    The number of vertices in the graph
    * @param  number_of_edges       The number of edges in the graph
-   * @param  has_data              Wiether or not the class has data, default = False
+   * @param  has_data              Wether or not the class has data, default = False
```
Comment (divyegala) [inline cpp/include/graph.hpp:389]: `Whether`

File: cpp/include/graph.hpp [hunk 2]
```diff
@@ -416,6 +416,14 @@ class GraphCOO {
         rmm::device_buffer{graph.edge_data, graph.number_of_edges * sizeof(weight_t), stream, mr};
     }
   }
+  GraphCOO(GraphCOOContents<vertex_t, edge_t, weight_t> &&contents)
+    : number_of_vertices_p(contents.number_of_vertices),
+      number_of_edges_p(contents.number_of_edges),
+      src_indices_p(std::move(*contents.src_indices.release())),
```
Comment (divyegala) [inline cpp/include/graph.hpp:422]: Maybe `*(contents.src_indices.release())` for readability to tell us the de-ref operator is on the pointer returned by the call to `release()`

File: python/cugraph/tests/test_maximum_spanning_tree.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+import warnings
+
+with warnings.catch_warnings():
+    warnings.filterwarnings("ignore", category=DeprecationWarning)
+    import networkx as nx
+
+print("Networkx version : {} ".format(nx.__version__))
+
+
+def compare_mst(mst_cugraph, mst_nx):
```
Comment (divyegala) [inline python/cugraph/tests/test_maximum_spanning_tree.py:42]: Duplicated with `test_minimum_spanning_tree.py`. Ideally it should be moved to a utils file shared by the two, or one file can host the tests and the other file can import

File: python/cugraph/tests/test_maximum_spanning_tree.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    # check total weight
+    cg_sum = edgelist_df["weights"].sum()
+    nx_sum = mst_nx_df["weight"].sum()
+    print(cg_sum)
+    print(nx_sum)
+    assert np.isclose(cg_sum, nx_sum)
+
+
+@pytest.mark.parametrize("graph_file", utils.DATASETS_UNDIRECTED_WEIGHTS)
+def test_maximum_spanning_tree_nx(graph_file):
```
Comment (divyegala) [inline python/cugraph/tests/test_maximum_spanning_tree.py:69]: This and the below test should be combined with the tests from `test_minimum_spanning_tree.py`. Tests can take parameter like:
```
@pytest.mark.parametrize("spanning_tree_class", ["maximum", "minimum"])
```
Then, if it's type maximum, then just negate the weights

File: python/cugraph/tree/minimum_spanning_tree_wrapper.pyx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    if not input_graph.adjlist:
+        input_graph.view_adj_list()
+    [offsets, indices] = graph_primtypes_wrapper.datatype_cast([input_graph.adjlist.offsets, input_graph.adjlist.indices], [np.int32])
+    num_verts = input_graph.number_of_vertices()
+    num_edges = input_graph.number_of_edges(directed_edges=True)
+
+    if input_graph.adjlist.weights is not None:
+        [weights] = graph_primtypes_wrapper.datatype_cast([input_graph.adjlist.weights], [np.float32, np.float64])
+    else:
+        weights = cudf.Series(np.full(num_edges, 1.0, dtype=np.float32))
```
Comment (divyegala) [inline python/cugraph/tree/minimum_spanning_tree_wrapper.pyx:63]: Does cugraph have a dependency on cupy? `cp.full` will likely be better here

File: python/cugraph/tree/minimum_spanning_tree_wrapper.pyx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
```
Comment (divyegala) [inline python/cugraph/tree/minimum_spanning_tree_wrapper.pyx:81]: Same as above

**Review by divyegala — CHANGES_REQUESTED** [review]
Thanks, Alex! Suggesting some changes and leaving a note for print statements in tests that could be commented/deleted before final review

File: cpp/include/graph.hpp (hunk 1 above)
Comment (afender) [reply cpp/include/graph.hpp]: third time the charm :) 0c6b9eba

File: cpp/include/graph.hpp (hunk 2 above)
Comment (afender) [reply cpp/include/graph.hpp]: Added 0c6b9eba

File: python/cugraph/tests/test_maximum_spanning_tree.py (hunk 3 above)
Comment (afender) [reply python/cugraph/tests/test_maximum_spanning_tree.py]: Moved to test/utils.py 0c6b9eba

File: python/cugraph/tests/test_maximum_spanning_tree.py (hunk 4 above)
Comment (afender) [reply python/cugraph/tests/test_maximum_spanning_tree.py:40]: Not sure about that. The goal here is to test `maximum_spanning_tree` feature. 
cuGraph does negate the weights itself actually as part of the feature but doing it manually in the test would defeat the purpose of testing the feature. Also, we need to compare against Nx's `maximum_spanning_tree`. As a result, notice that merging would result in branches all over in the tests.

File: python/cugraph/tree/minimum_spanning_tree_wrapper.pyx (hunk 5 above)
Comment (afender) [reply python/cugraph/tree/minimum_spanning_tree_wrapper.pyx]: It was just added by Rick actually. Updated 0c6b9eba

File: python/cugraph/tree/minimum_spanning_tree_wrapper.pyx (hunk 6 above)
Comment (afender) [reply python/cugraph/tree/minimum_spanning_tree_wrapper.pyx]: fixed 0c6b9eba

**Review by divyegala — APPROVED** [review]
LGTM, thank Alex!

**Review by BradReesWork — APPROVED** [review]

**Pull request merged**